python_modules/dagster-graphql/dagster_graphql/implementation/pipeline_run_storage.py

```diff
@@ -1,27 +1,83 @@
+from enum import Enum
+from dagster import check
+from threading import Event, Thread
+from time import sleep
+
+import gevent
+
+
+class State(Enum):
+    # initialized
+    NULL = 0
+    # streaming the existing event log entries
+    LOADING = 1
+    # watching for new writes
+    WATCHING = 2
+
+
+CHUNK_SIZE = 10000
+
+
 class PipelineRunObservableSubscribe:
     def __init__(self, instance, run_id, after_cursor=None):
         self.instance = instance
         self.run_id = run_id
         self.observer = None
+        self.state = State.NULL
+        self.load_thread = None
+        self.stopping = None
         self.after_cursor = after_cursor if after_cursor is not None else -1
 
     def __call__(self, observer):
         self.observer = observer
+        check.invariant(self.state is State.NULL, f"unexpected state {self.state}")
+        self.state = State.LOADING
 
-        events = self.instance.logs_after(self.run_id, self.after_cursor)
-        if events:
-            self.observer.on_next(events)
+        # support for gevent based dagit server
+        if gevent.get_hub().gr_frame:
+            self.stopping = gevent.event.Event()
+            self.load_thread = gevent.Greenlet(self.load_events, gevent.sleep)
+        else:
+            self.stopping = Event()
+            self.load_thread = Thread(
+                target=self.load_events,
+                args=(sleep,),
+                name=f"load-events-{self.run_id}",
```

Comment:
Does this name have to be unique?  What happens if I open the same run on multiple tabs?

Reply:
https://docs.python.org/3/library/threading.html#threading.Thread.name